### PR TITLE
[Core]Fix placement group gpuids assigned issue

### PR DESCRIPTION
**Pull request by clay4megtr**
## Why are these changes needed?

Currently, the data struct `ResourceInstanceCapacities` represent a resource type, like `CPU`  `GPU`  `CPU_Pg_xxx` and so on, and it's in order, for `GPU`, the order index represents which card is it, before this pr, when we commit the bundle resources  into cluster resource scheduler, we add bundle commit resource directly and ignored the index order.

## Related issue number
#14759 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/placement_group_resource_manager_test.cc [hunk 1]
```diff
@@ -154,7 +154,8 @@ TEST_F(NewPlacementGroupResourceManagerTest, TestNewMultipleBundlesCommitAndRetu
       std::make_shared<TaskResourceInstances>();
   ASSERT_TRUE(remaining_resource_scheduler->AllocateLocalTaskResources(
       init_unit_resource, resource_instances));
-  auto remaining_resouece_instance = remaining_resource_scheduler->GetLocalResources();
+  auto remaining_resouece_instance = remaining_resource_scheduler->GetLocalNodeResources();
```
Comment (richardliaw) [inline src/ray/raylet/placement_group_resource_manager_test.cc:157]: can we fix this typo in this PR? `remaining_resouece_instance` -> `remaining_resource_instance

**Comment by rkooo567** [comment]
Can you update the problem and fix in the description?

**Comment by wuisawesome** [comment]
+1 to including a description of the bug/solution. I don't understand why the cluster resource scheduler needs to know about the concept of a bundle.

File: src/ray/raylet/placement_group_resource_manager_test.cc (hunk 1 above)
Comment (clay4444) [reply src/ray/raylet/placement_group_resource_manager_test.cc:157]: sure,  thanks for the remaining.

**Comment by clay4444** [comment]
> +1 to including a description of the bug/solution. I don't understand why the cluster resource scheduler needs to know about the concept of a bundle.

yeah, that's a good point, the concept of the bundle shouldn't be involved in the cluster resource scheduler. I will update this

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 2]
```diff
@@ -404,6 +404,57 @@ const NodeResources &ClusterResourceScheduler::GetLocalNodeResources() const {
...
+    const std::string &predefined_resource_name,
+    const std::string &customer_resource_name, const double &customer_resource_value,
+    const std::shared_ptr<TaskResourceInstances> resource_instances) {
+  RAY_CHECK(resource_instances->custom_resources.size() == 0);
+
+  string_to_int_map_.Insert(customer_resource_name);
+  int64_t resource_id = string_to_int_map_.Get(customer_resource_name);
+
+  int idx = -1;
+  if (predefined_resource_name == ray::kCPU_ResourceLabel) {
```
Comment (ffbin) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:417]: Switch case is better?

**Comment by wuisawesome** [comment]
I think I understand the bug now, but shouldn't the solution be simpler than this? 

Right now we call `void AddLocalResource(const std::string &resource_name, double resource_total);`

Shouldn't we just change that function to `void AddLocalResourceInstances(const std::string &resource_name, const TaskResourceInstances&);`?

**Comment by clay4444** [comment]
> I think I understand the bug now, but shouldn't the solution be simpler than this?
> 
> Right now we call `void AddLocalResource(const std::string &resource_name, double resource_total);`
> 
> Shouldn't we just change that function to `void AddLocalResourceInstances(const std::string &resource_name, const TaskResourceInstances&);`?

hmm… but I think these two arguments `predefined_resource_name` and `customer_resource_value` are necessary, right?  for example if we don't specify `predefined_resource_name`, how do we know which original resource we want to commit?  and also, from naming, this operation of commit bundle is not only an added resource operation but also a `transform` operation, it will transform the predefined resource into a new customer resource, so I think the `TransformResource` would be better.

**Comment by wuisawesome** [comment]
Why do you need to know the original resource you're committing? The allocation already happened during the prepare step, so you should never need to modify that resource again. You also don't need to read it, because you already store the allocated instances with the bundle.

**Comment by clay4444** [comment]
> Why do you need to know the original resource you're committing? The allocation already happened during the prepare step, so you should never need to modify that resource again. You also don't need to read it, because you already store the allocated instances with the bundle.

actually, the reason why we need the original resource when commit resource is we prepared(allocate) the `ORIGINAL RESOURCE` in the prepared phase, but commit(add) the bundle customer resource(like "CPU_PG_xx") int the commit phase.

you can imagine this case: we commit the resource `CPU_PG_1` and the TaskResourceInstance is [ [0,1], [1,0] ],  we can't  make sure we should commit [0,1] or should commit [1, 0],

**Comment by rkooo567** [comment]
`test_global_state` seems to fail.

**Comment by rkooo567** [comment]
Also, I don't quite understand what this means; 
```
we add bundle commit resource directly and ignored the index order.
```

Can you show me an example? (with indexes)

**Comment by clay4444** [comment]
> Also, I don't quite understand what this means;
> 
> ```
> we add bundle commit resource directly and ignored the index order.
> ```
> 
> Can you show me an example? (with indexes)

sure, you can imagine this case:  there are two bundle resource which will be committed is <GPU_group_1_xxx, 1.0>, <GPU_group_2_xxx, 1.0>  and the allocated resource (which  type is `TaskResourceInstance`) is [ [0,0], [0,1] ] and [ [0,1] [1,0] ], the second vector represent the `GPU` resource, the commit result of these two bundles are both `[ [0,0], [1,0] ]` because we ignored the index order directly.

**Comment by clay4444** [comment]
hi @wuisawesome @rkooo567   I think this pr can be merged?

**Comment by wuisawesome** [comment]
I still don't see the problem.

==================

In your example, after the prepare phase is over, before any commits occur, the state of the node is:

`{"GPU": [0, 0]}`

=================

After the first bundle is committed, we call 

```
AddLocalResourceInstance( "<GPU_group_1_xxx, 1.0>", [1, 0]);
```

Now the node state is `{"GPU": [0, 0], "<GPU_group_1_xxx, 1.0>": [1, 0]}`

==================

Now the second bundle is committed

 ```
AddLocalResourceInstance( "<GPU_group_2_xxx, 1.0>", [0, 1]);
```

Now the node state is `{"GPU": [0, 0], "<GPU_group_1_xxx, 1.0>": [1, 0], "<GPU_group_2_xxx, 1.0>": [0, 1]}`
==================

**Comment by wuisawesome** [comment]
Can you look at #15049?

**Pull request closed without merging**